### PR TITLE
Enable calling caffe2 LayerNorm from PyTorch and JIT

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#15243 Enable calling caffe2 LayerNorm from PyTorch and JIT**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13473791/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15312 Remove Context from c10 operator schemas&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13495323/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15316 Move files to/from c10/core and c10/util&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13498493/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15317 Clean up Half&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13498492/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15407 caffe2::Tensor::is_same()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13520389/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15853 Move blob to c10&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605256/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15854 Code style cleanup&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605258/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15855 Remove some dependencies from ivalue.h to ATen&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605259/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15856 Fix export macros&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605257/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15941 Split IValue -> Tensor dependency, move to c10&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13630156/)

Register it as a custom JIT op.

Differential Revision: [D13473791](https://our.internmc.facebook.com/intern/diff/D13473791/)

### CODE REVIEW

File: caffe2/operators/layer_norm_op.cc
Comment (smessmer) [reply caffe2/operators/layer_norm_op.cc]: I agree it's confusing. It was `Tensor*` originally, but this actually also didn't really make sense because Tensors don't respect constness anyhow. As long as we don't have `ConstTensor` or something like that, I don't think there's a good solution.

About pointer vs const-ref: when it's a pointer, it needs to live on the stack somewhere, i.e. when calling the op from caffe2 with a caffe2::Tensor, I need to create a vector<C10Tensor> for the outputs, which is an unnecessary additional allocation on each call. Passing it by value or const-ref allows me to just wrap each caffe2::Tensor into a C10Tensor constructor on call site, avoiding the vector.

I'm not sure what you mean with "magic-point" and "alloc hint", can you elaborate?

But let's actually not block this PR on that discussion. I prefer landing this now and changing it later.

File: torch/csrc/jit/c10_ops/layer_norm.cpp
Comment (smessmer) [reply torch/csrc/jit/c10_ops/layer_norm.cpp]: I'm planning to do that in a separate PR.

File: torch/csrc/jit/c10_ops/layer_norm.cpp
Comment (ezyang) [reply torch/csrc/jit/c10_ops/layer_norm.cpp]: Yeah, this is wrong. If you do this, you'll generate a leaf variable that doesn't require grad. This will lead to hard to debug problems if someone tries to differentiate through this, because the gradients will just stop at the outputs of this op and accumulate there.

Comment (ezyang) [reply torch/csrc/jit/c10_ops/layer_norm.cpp]: Also CC @goldsborough because I know autograd support for custom operators is not done yet, and I don't know what the eventual design was planned to be.

**Review by ezyang — CHANGES_REQUESTED** [review]
There are no tests

File: torch/csrc/jit/c10_ops/layer_norm.cpp
Comment (smessmer) [reply torch/csrc/jit/c10_ops/layer_norm.cpp]: Afaik, this will always get a variable and should always return a variable. However, this variable might have `requires_grad==true` or `requires_grad==false`. This code only handles `requires_grad=false` correctly, that's why there is an exception thrown in line 16 if `requires_grad==true`.

Comment (ezyang) [reply torch/csrc/jit/c10_ops/layer_norm.cpp]: OK... that seems convincing enough.

Still, all of this boilerplate seems like a massive antipattern. The way you have structured the code here, seems to imply that everyone who wants to write an operator this way has to write the boilerplate of unwrapping and then rewrapping the variable. This is why we had code generation in the old version of this code. If you are going to say that, "Ah, I'll figure out how to solve this problem later," it would be nice if this code, somewhere, acknowledged the issue, rather than having no comments at all. But I am not sure at all how you are going to solve this problem; we talked about a boxed representation so that we can uniformly perform common operator operations, including wrapping and unwrapping, but I am afraid that it has passed out of the collective memories of everyone. Doc is here https://fb.quip.com/o5pXA1LHgAq0 or in this Gist for OSS people following: https://gist.github.com/ezyang/0ff02f92d742c2587fd76b640eef9e11 )

Also cc @yf225 because all of this changes when Tensor and Variable merge in C++ (but this doesn't "solve" your problem, because you still need to deal with requires grad = true).

Comment (smessmer) [reply torch/csrc/jit/c10_ops/layer_norm.cpp]: I added a detailed comment.

**Comment by smessmer** [comment]
Tests will be added in a separate PR

**Comment by smessmer** [comment]
correction: will add tests here before landing

**Pull request closed without merging**
